### ISSUE TITLE
Inter-test dependencies

### ISSUE THREAD

**Issue by feuerbach** [feuerbach/tasty#48]
@jstolarek writes:

> I want to run test A first and if it succeeds I want to run test B (if A fails B is skipped).

@joeyh has also requested this feature in #47.

The main problem here is a syntax/API for declaring dependencies. It would be relatively easy for hspec-style monadic test suites, but I don't see a good way to do it using the current list-based interface.

(This could be a reason to move to the monadic interface. But it's a big change, and I'd like to explore other options first.)

**Comment by mightybyte** [comment]
I think I agree with @ocharles that the monadic interface feels like the way to go here.  It seems like we need the most powerful abstraction possible at the lowest level, otherwise we keep end up running into these situations where as [Ed says](https://www.youtube.com/watch?v=yFXzuCFeRGM&feature=youtu.be&t=1h34m55s), "we can't say all the things".

**Comment by quchen** [comment]
Bump because I’ve been missing this feature a lot for a long time now!

```haskell
onSuccessRun :: TestTree -> TestTree -> TestTree
onSuccessRun precondition moreTests = ...
```

**Comment by feuerbach** [comment]
@quchen as I said above,

> I don't have resources to work on this atm (and in the foreseeable future), but you can totally try it yourself. I'm happy to review your early designs/prototypes.

The API you propose looks fine to me.

**Comment by quchen** [comment]
Oh, I missed that. But it’s good to know you’re open to such a feature, so contributors can be sure it is merged once they get it done. :-)

**Comment by feuerbach** [comment]
I've been slowly working on this feature, and now it's pretty much finished (except documentation) in the `dependencies` branch.

With it, you can apply `after AllFinish "pattern"` or `after AllSucceed "pattern"` to a test tree, and it will make all the tests inside the tree execute only after all tests whose name match the pattern have finished (respectively, succeeded).

Here's a minimal example:

``` haskell
{-# LANGUAGE NumDecimals #-}
import Test.Tasty
import Test.Tasty.HUnit
import Control.Concurrent

main = defaultMain $ testGroup "Tests" $
  [ after AllSucceed "Three" $ testCase "One" $ threadDelay 1e6
  , testCase "Two" $ threadDelay 1e6
  , testCase "Three" $ threadDelay 1e6 >> assertFailure "fail"
  ]
```

Let me know what you think.

**Comment by RyanGlScott** [comment]
This looks extremely cool, @feuerbach. I'm eager to incorporate this feature into the `singletons` test suite (which would greatly benefit from this feature).

I've been playing around with the `dependencies` branch for a bit and trying out some toy examples. One thing that's unclear to me (and perhaps you intend to document this) is what happens if you declare recursive dependencies, such as in the following example:

```haskell
{-# LANGUAGE NumDecimals #-}
import Test.Tasty
import Test.Tasty.HUnit
import Control.Concurrent

main :: IO ()
main = defaultMain $ testGroup "Tests" $
  [ after AllSucceed "Foo" $ testCase "Foo" $ threadDelay 1e6
  ]
```

Currently, this loops infinitely at runtime (instead of, say, throwing an error). Is this the intended behavior?

**Comment by feuerbach** [comment]
Good catch Ryan, should be fixed now.

**Comment by RyanGlScott** [comment]
Thanks @feuerbach!

I've encountered another awkward design consideration that I'm not sure how to address. [Here](https://github.com/goldfirere/singletons/blob/74176fa8e8392f21e116dd2b4b259aed2076d7c9/tests/SingletonsTestSuite.hs#L115-L117) is some code from the `singletons` test suite (simplified for presentation purposes):

```haskell
testGroup "Database client"
    [ testCase "Database" $ threadDelay 1e6
    , testCase "Main" $ threadDelay 1e6
    ]
```

I want the `Main` test case to run only after the `Database` test case has finished. I attempted this using `after` like so:

```haskell
testGroup "Database client" $
    [ testCase "Database" $ threadDelay 1e6
    , after AllFinish "Database" $
      testCase "Main" $ threadDelay 1e6
    ]
```

Unfortunately, this doesn't quite do what I would have hoped:

```
λ> main
*** Exception: Test dependencies form a loop.
```

After some head-scratching, I realized that the `Database` pattern was matching the entire `Database client` test group, which resulted in cyclic dependencies. On the other hand, this wasn't my intention—I wanted to match the `Database` test case in particular. Is there a way that I can communicate this fact to `tasty`?

**Comment by feuerbach** [comment]
Yes, you can do that since tasty patterns are awk expressions. Instead of `"Database"`, try something like `"$3 == \"Database\""`, where 3 is the level at which that test group occurs in the tree (the outermost group is number 1). See https://github.com/feuerbach/tasty/blob/master/README.md#patterns for the details.

**Comment by RyanGlScott** [comment]
Ah, I completely missed that, thanks.

A design question: what is the best idiom for depending on multiple tests? Here are two semantically equivalent `TestGroup`s:

```haskell
main :: IO ()
main = defaultMain $
  testGroup "Tests" $
  [ testGroup "Group1"
    [ after AllSucceed "Bar" $
      after AllSucceed "Baz" $
      testCase "Foo" $ threadDelay 1e6
    ]
  , testGroup "Group2"
    [ testCase "Bar" $ threadDelay 1e6
    , testCase "Baz" $ threadDelay 1e6
    ]
  ]
```
```haskell
main :: IO ()
main = defaultMain $
  testGroup "Tests" $
  [ testGroup "Group1"
    [ after AllSucceed "$3 == \"Bar\" || $3 == \"Baz\"" $
      testCase "Foo" $ threadDelay 1e6
    ]
  , testGroup "Group2"
    [ testCase "Bar" $ threadDelay 1e6
    , testCase "Baz" $ threadDelay 1e6
    ]
  ]
```

I'm not sure if one of these idioms parallelizes better than the other, though. (I could imagine that stacking uses of `after` might affect how things are scheduled.) Do you suspect this to be the case? If so, do you think `after` should take a list of patterns instead of just a single one?

**Comment by feuerbach** [comment]
There's no semantic difference between the two styles—in both cases parallelism is exploited to the maximum extent allowed by the dependencies—so feel free to use the style you find more readable.

There is however a semantic difference between `"Bar"` and `"$3 == \"Bar\"`. A more accurate translation of your first example would be `/Bar/ || /Baz/`.

**Comment by RyanGlScott** [comment]
Good to know, thanks. I was able to port over the entirety of the `singletons` test suite to express all of its test dependencies. Now I can finally run those tests in parallel! (On my machine using four threads, this reduces the total run time from about 90 seconds to 30 seconds, which is pretty darn cool.)

I suppose if I had one final comment to make about the proposed API, it's that I had to be extremely cautious when typing in the patterns it depended on, since it's possible to "depend" on tests which don't exist. For example:

```haskell
main :: IO ()
main = defaultMain $
  testGroup "Tests" $
  [ after AllSucceed "Baz" $
    testCase "Foo" $ threadDelay 1e6
  , testCase "Bar" $ threadDelay 1e6
  ]
```

The `Foo` test case runs without issue, despite the fact that no tests match the `Baz` pattern. I suppose one could argue that this makes sense since `Foo` did in fact wait for all of its dependencies to succeed (all zero of them) before it ran. On the other hand, it does mean that if one were to change the name of a test which had reverse dependencies and forgot to update the uses of `after` which depended on that test's old name, then one might not realize their mistake.

I wouldn't be too bothered either way on this point, but I thought it was worth pointing out.

**Comment by martijnbastiaan** [comment]
Thanks for the hard work @feuerbach ! It's really nice to see this being addressed. 

Unfortunately.. I'm having some issues incorporating this into the code I'm working on. Here's the context: I'm currently working on a Haskell to hardware compiler. To test the compiler we compile some designs to a number of target languages and run simulators on those languages to check if they have some expected output. A typical test case for one design looks something like:

```
I2C:
    VHDL:
        haskell-to-vhdl
        vhdl-compile
        vhdl-simulation
    Verilog:
        haskell-to-verilog
        verilog-compile
        verilog-simulation
    ......
```

where `I2C` is the program/circuit we're compiling, `VHDL` and `Verilog` are target languages, and the most inner steps are the actual test. We'd like to keep this granularity, so I previously hacked together a solution with locks and unsafe IO to make the inner tests run sequentially. In reality, we don't specify all these testcases manually. Instead, we only specify it on an `I2C` level and the *actual* tests are generated, as it would be tedious to manually add tests for `vhdl/verilog/..` everytime. The code generating these tests ideally would also make the tests dependent on each other, such that they run sequentially. 

Now.. I've actually written this code. For the following test cases:

```
Tests.Examples.I2C.VHDL.clash
Tests.Examples.I2C.VHDL.GHDL (library) [test_bitmaster]
Tests.Examples.I2C.VHDL.GHDL (library) [test_bytemaster]
```

It should run `clash`, then `test_bitmaster`, then `test_bytemaster`. So, my code generates the following patterns:

| TestName                         | Pattern                                                                                     |
|----------------------------------|---------------------------------------------------------------------------------------------|
| clash                            |                                                                                             |
| GHDL (library) [test_bitmaster]  | ($(NF-0) == "clash") && ($(NF-1) == "VHDL") && ($(NF-2) == "I2C")                           |
| GHDL (library) [test_bytemaster] | ($(NF-0) == "GHDL (library) [test_bitmaster]") && ($(NF-1) == "VHDL") && ($(NF-2) == "I2C") |

But once I actually run the tests, the run in parallel! I tried changing the patterns to:

| TestName                         | Pattern                                                               |
|----------------------------------|-----------------------------------------------------------------------|
| clash                            |                                                                       |
| GHDL (library) [test_bitmaster]  | ($(NF-0) == "clash") && ($(NF-1) == "VHDL")                           |
| GHDL (library) [test_bytemaster] | ($(NF-0) == "GHDL (library) [test_bitmaster]") && ($(NF-1) == "VHDL") |

At this point it the tests run sequentially as they should, but there's no need for `GHDL (library) [test_bitmaster]` to wait for *all* the other clash compilations! I've also tried the following patterns:

| TestName                         | Pattern                                                              |
|----------------------------------|----------------------------------------------------------------------|
| clash                            |                                                                      |
| GHDL (library) [test_bitmaster]  | ($(NF-0) == "clash") && ($(NF-2) == "I2C")                           |
| GHDL (library) [test_bytemaster] | ($(NF-0) == "GHDL (library) [test_bitmaster]") && ($(NF-2) == "I2C") |

which again runs in parallel. It seems to come down to `$(NF-2)`, which just doesn't seem to work in dependent tests for some reason. Interestingly enough, if I use it as a pattern on the command line (with `-p`), they do match as I would expect!

Would you have any idea on where this might go wrong? Am I using the API incorrectly?

**Comment by feuerbach** [comment]
Thanks for trying this out, Martijn. I struggle a bit to understand what your test tree looks like. Could you try to create a minimal complete example that demonstrates your issue?

**Comment by martijnbastiaan** [comment]
Yes, I should have done that in the first place. Setup:

```haskell
module Main (main) where

import Control.Concurrent (threadDelay)

import Test.Tasty
import Test.Tasty.Clash
import Test.Tasty.HUnit

acquire :: IO ()
acquire = return ()

release :: a -> IO ()
release _ = return ()
```

The following executes as expected:

```haskell
main :: IO ()
main = do
  defaultMain $ testGroup "L1"
    [ testGroup "L2"
      [ testCase "L2A" (threadDelay 1000000)
      , after AllFinish "($(NF-0) == \"L2A\") && ($(NF-1) == \"L2\") && ($(NF-2) == \"L1\")" $
          testCase "L2B" (threadDelay 1000000)
      ]
    ]
```

output:

```
L1
  L2
    L2A: OK (1.00s)
    L2B: OK (1.00s)

All 2 tests passed (2.00s)
```

Notice that it takes 2 seconds. Now the following:

```haskell
main :: IO ()
main = do
  defaultMain $ testGroup "L1"
    [ withResource acquire release $ const $ testGroup "L2"
      [ testCase "L2A" (threadDelay 1000000)
      , after AllFinish "($(NF-0) == \"L2A\") && ($(NF-1) == \"L2\") && ($(NF-2) == \"L1\")" $
          testCase "L2B" (threadDelay 1000000)
      ]
    ]
```

output:

```
L1
  L2
    L2A: OK (1.00s)
    L2B: OK (1.00s)

All 2 tests passed (1.00s)
```

So maybe I'm simply using `withResource` incorrectly? It is supposed to setup some infrastructure (basically a number of `mkdir`s) and clean it out after the test finished (basically `rm -rf`).

**Comment by martijnbastiaan** [comment]
Concerning the discussion about `after`: I think it should produce an error if it matched zero tests. It's hard to imagine a situation where you'd want to match zero tests, and I've been bitten by this multiple times now. Of course, some people might still be interested in matching zero or more tests, so an additional `after0` could be added. 

> One solution I considered was a version of the after combinator which would take the expected number of the tests matching the pattern, and it would error if the wrong number of tests matched the pattern.

That would be great! In my case, the code generating the patterns knows perfectly well how many cases should be matched. Having this feature would decrease the possible number of name collisions in our testsuite to zero :). Though it must be said that this falls squarely into the category 'nice to have', not 'need to have' for my case.

**Comment by feuerbach** [comment]
> I think it should produce an error if it matched zero tests

I think so too; unfortunately, it's a bit complicated to implement. The issue is that tests may be filtered, and when implemented naively, the error will trigger when the dependency is not included in the set of running tests. Implementing this properly requires some inconvenient bookkeeping. But I might get around to it some day.

**Comment by martijnbastiaan** [comment]
That makes sense. Maybe I'll take a shot at it in a few weekends from now.

In case you missed it due to my second comment: I posted a minimal example of the bug (?) I'm experiencing. I've looked through the code added in the `dependencies` branch, but I haven't seen any obvious implementation mistakes yet.

**Comment by feuerbach** [comment]
Yes, I saw your example (thanks!) and can reproduce it — looks like a bug. I'll look into it.

**Comment by martijnbastiaan** [comment]
I've done some digging. `Test.Tasty.Pattern.exprMatches` does not get the complete path sometimes. If the call comes from `testPatternMatches` it does. This leaves only one option as `exprMatches` is used in only one other place: `resolveDeps`. `resolveDeps` doesn't do anything funny with the path though, it just gets them from `createTestActions`. As we're only seeing the bug in combination with `withResource`, a likely source of the error is `addInitAndRelease` (use for `foldResource`) which might forget to pass the path-so-far along:

```haskell
    addInitAndRelease :: ResourceSpec a -> (IO a -> Tr) -> Tr
    addInitAndRelease (ResourceSpec doInit doRelease) a = wrap $ do
      initVar <- atomically $ newTVar NotCreated
      (tests, fins) <- unwrap $ a (getResource initVar)
      let ntests = length tests
      finishVar <- atomically $ newTVar ntests
      let
        ini = Initializer doInit initVar
        fin = Finalizer doRelease initVar finishVar
        tests' = map (first $ local $ (Seq.|> ini) *** (fin Seq.<|)) tests
      return (tests', fins Seq.|> fin)
```

This is speculation though, as I'm having a hard time tracing/understanding due to all the abstractions going on.

**Comment by martijnbastiaan** [comment]
This patch solves my problems:

```diff
diff --git a/core/Test/Tasty/Core.hs b/core/Test/Tasty/Core.hs
index 812e4f5..df75042 100644
--- a/core/Test/Tasty/Core.hs
+++ b/core/Test/Tasty/Core.hs
@@ -292,7 +292,7 @@ after deptype s =
 data TreeFold b = TreeFold
   { foldSingle :: forall t . IsTest t => OptionSet -> TestName -> t -> b
   , foldGroup :: TestName -> b -> b
-  , foldResource :: forall a . ResourceSpec a -> (IO a -> b) -> b
+  , foldResource :: forall a . ResourceSpec a -> Path -> (IO a -> b) -> b
   , foldAfter :: DependencyType -> Expr -> b -> b
   }
 
@@ -311,7 +311,7 @@ trivialFold :: Monoid b => TreeFold b
 trivialFold = TreeFold
   { foldSingle = \_ _ _ -> mempty
   , foldGroup = const id
-  , foldResource = \_ f -> f $ throwIO NotRunningTests
+  , foldResource = \_ _ f -> f $ throwIO NotRunningTests
   , foldAfter = \_ _ b -> b
   }
 
@@ -355,7 +355,7 @@ foldTestTree (TreeFold fTest fGroup fResource fAfter) opts0 tree0 =
         TestGroup name trees ->
           fGroup name $ foldMap (go pat (path Seq.|> name) opts) trees
         PlusTestOptions f tree -> go pat path (f opts) tree
-        WithResource res0 tree -> fResource res0 $ \res -> go pat path opts (tree res)
+        WithResource res0 tree -> fResource res0 path $ \res -> go pat path opts (tree res)
         AskOptions f -> go pat path opts (f opts)
         After deptype dep tree -> fAfter deptype dep $ go pat path opts tree
 
diff --git a/core/Test/Tasty/Run.hs b/core/Test/Tasty/Run.hs
index 27f302e..b564ce9 100644
--- a/core/Test/Tasty/Run.hs
+++ b/core/Test/Tasty/Run.hs
@@ -242,7 +242,7 @@ createTestActions opts0 tree = do
               Traversal $ local (second ((deptype, pat) :)) a
           }
         opts0 tree
-  (tests, fins) <- unwrap traversal
+  (tests, fins) <- unwrap mempty traversal
   let
     mb_tests :: Maybe [(Action, TVar Status)]
     mb_tests = resolveDeps $ map
@@ -263,10 +263,10 @@ createTestActions opts0 tree = do
         act (inits, fins) =
           executeTest (run opts test) statusVar (lookupOption opts) inits fins
       tell ([(act, (statusVar, path, deps))], mempty)
-    addInitAndRelease :: ResourceSpec a -> (IO a -> Tr) -> Tr
-    addInitAndRelease (ResourceSpec doInit doRelease) a = wrap $ do
+    addInitAndRelease :: ResourceSpec a -> Path -> (IO a -> Tr) -> Tr
+    addInitAndRelease (ResourceSpec doInit doRelease) p a = wrap $ do
       initVar <- atomically $ newTVar NotCreated
-      (tests, fins) <- unwrap $ a (getResource initVar)
+      (tests, fins) <- unwrap p $ a (getResource initVar)
       let ntests = length tests
       finishVar <- atomically $ newTVar ntests
       let
@@ -279,9 +279,10 @@ createTestActions opts0 tree = do
       -> Tr
     wrap = Traversal . WriterT . lift . fmap ((,) ())
     unwrap
-      :: Tr
+      :: Path
+      -> Tr
       -> IO ([(InitFinPair -> IO (), (TVar Status, Path, Deps))], Seq.Seq Finalizer)
-    unwrap = flip runReaderT mempty . execWriterT . getTraversal
+    unwrap p = flip runReaderT (p, mempty) . execWriterT . getTraversal
 
 -- | Take care of the dependencies.
 --
```

I'm unsure if this is the way to go. I'll see if I can find any issues with it.

**Comment by martijnbastiaan** [comment]
At the risk of spamming.. you can test this patch by:

```bash
git remote add martijnbastiaan https://github.com/martijnbastiaan/tasty.git
git fetch martijnbastiaan
git cherry-pick 0d2dee1341a94ce22d17f12165528e8975c88bce
git remote remove martijnbastiaan
```

**Comment by feuerbach** [comment]
Thanks Martijn, the fix wasn't quite right but your diagnosis saved me a lot of time. This should be fixed now.

**Comment by martijnbastiaan** [comment]
Neat, thanks Roman! I've successfully implemented this feature in the Clash testsuite, and it works really well. Looking forward to the next Tasty release :-).

**Comment by feuerbach** [comment]
Dependencies are now [documented](https://github.com/feuerbach/tasty#dependencies) in the README and merged into master.

I'll wait for another week or two for any remaining feedback/bug reports and then make a release.

**Comment by feuerbach** [comment]
Ok, interesting.

One thing I'd suggest is to construct the pattern AST directly instead of assembling a string pattern. You can see an example of that here: https://github.com/feuerbach/tasty/blob/master/benchmarks/dependencies/test.hs#L37-L38

**Comment by martijnbastiaan** [comment]
Right, that's better. Thanks :+1: .

**Comment by feuerbach** [comment]
[This is now released as part of tasty-1.2][ann].

[ann]: https://ro-che.info/articles/2018-12-01-tasty-dependencies
